activerecord/lib/active_record/middleware/strict_safe_http_methods.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+# :markup: markdown
+
+module ActiveRecord
+  module Middleware
+    # # Strict Safe HTTP Methods \Middleware
+    #
+    # This middleware prevents database writes while serving a "safe" HTTP request (`GET`, `HEAD`, or
+    # `OPTIONS`) by raising an `ActiveRecord::ReadOnlyError` exception if a database write is
+    # attempted.
+    #
+    # Disabled by default, this middleware can be added by setting
+    # `config.active_record.strict_safe_http_methods` to `true`.
+    class StrictSafeHTTPMethods
+      def initialize(app)
+        @app = app
+      end
+
+      def call(env)
+        request = ActionDispatch::Request.new(env)
```

Comment:
I'm not convinced this middleware should be in Active Record. 

Logic that depends on two frameworks are always a bit tricky, but I think it would makes more sense to be in `actionpack`, and would need to check for Active Record presence. A somewhat similar case is how Active Job defer push to after commit if Active Record is present.

Reply:
When I started drafting this change, I put it in Action Pack, but changed my mind after seeing how similar the ShardSelector and DatabaseSelector middlewares are (see [`default_middlware_stack.rb` in particular](https://github.com/rails/rails/blob/1d1790b04d3930d65c2436bbe1d381af821d0ef4/railties/lib/rails/application/default_middleware_stack.rb#L95-L107)). That said, I don't have a strong opinion and will try moving it back to see how it feels.